src/games/strategy/triplea/ai/proAI/ProAI.java

```diff
@@ -482,11 +490,11 @@ public boolean selectAttackSubs(final Territory unitTerritory) {
     final List<Unit> defenders = (List<Unit>) battle.getDefendingUnits();
     ProLogger.info(player.getName() + " checking sub attack in " + unitTerritory + ", attackers=" + attackers
         + ", defenders=" + defenders);
-    ProBattleUtils.setData(getGameData());
+    calc.setData(getGameData());
 
     // Calculate battle results
     final ProBattleResult result =
-        ProBattleUtils.calculateBattleResults(player, unitTerritory, attackers, defenders, new HashSet<Unit>(), true);
+        calc.calculateBattleResults(player, unitTerritory, attackers, defenders, new HashSet<Unit>(), true);
```

Comment:
Don't you have to copy it every time? The battle result calc is destructive of the game data object, right? So you need to copy it in everytime before calculateBattleResults, right?  Couldn't the calculate method perhaps do the check to see if the game data object would need to be copied, rather than having clients manage the calc for it.